### ISSUE TITLE
Some BagIt Exports throw 504 error

### ISSUE THREAD

**Issue by lalgee** [LibraryOfCongress/concordia#1503]
**What behavior did you observe? Please describe the bug**
When trying to do a BagIt Export for the James A. Garfield Diary campaign in production, the Bag does not download and instead a 504 error page loads.

**How can we reproduce the bug?**
1. Go to [James A. Garfield Diary campaign in Django admin](https://crowd.loc.gov/admin/concordia/campaign/25/change/)
2. Click Export Bagit 
3. See 504 error

**What is the expected behavior?**
When the BagIt Export button is clicked, the user should remain on the page and a .zip BagIt file should download.

**Additional context**
The issue does not occur in stage, where none of the Garfield items are completed.

@jkueloc was able to download a completed Garfield Bag successfully from her local copy of the dev environment. This data can be used for ETL prep, so this issue is no longer a blocker to ETL of the campaign.

**Comment by jkueloc** [comment]
@lalgee We have a work around for this issue. On 7/12 after restoring my local environment database I was able to successfully BagIt export the Garfield campaign.  @laurenseroka reviewed the structure of that export and ok'd it. 
It was not  result of the regex fix on #1501 but another issue. Probably a memory resource constraint (redis, memcache, django, postgres ...?) and a need for db indexes that postgressql will actually use.

**Comment by lalgee** [comment]
Makes sense that this is a memory/performance issue. Tested BagIt export for all current campaigns seemed to also slow down the site frontend.

The following campaigns also give a 504 timeout error for BagIt export:
- “My great mass of papers”: Correspondence of Theodore Roosevelt
- Rough Rider to Bull Moose: Letters to Theodore Roosevelt
- The Blackwells: An Extraordinary Family
- Anna E. Dickinson Papers

This isn't an immediate priority, so I'm going to bump to another sprint. We should possibly also reassign to @rabiloc since he's been working in export - @elainekamlley ?

**Comment by jkueloc** [comment]
After clicking the Garfield campaign 'export BagIt' - the production log shows WARNINGS on 4 different health checks (healthz) Tasks with "took too long to shut down and was killed." before issuing a an INFO msg "Creating bag for directory /tmp/garfieldb_xas4s3" and then issuing the same WARNING on GET for uri=/admin/concordia/campaign/exportBagIt/garfield clientproto=HTTP/1.1>.

However, the log also shows that the process is continuing - through completion at least through tagmannifest-sha512.txt - though the connection to the client is lost. - no msgs on zipping up.

The msgs that follow are 5 WARNING same as the beginning except on method=POST uri=/reserve-asset/600377.
The log excerpt I'm referencing is attached....

[log-events-viewer-result-Garfield-Prod-20210713.csv](https://github.com/LibraryOfCongress/concordia/files/6856484/log-events-viewer-result-Garfield-Prod-20210713.csv)

**Comment by jkueloc** [comment]
The 2 immediate SQL statements from the log in a local dev environment with a containerized database show the following:
[2021-07-14T13:15:41 DEBUG django.db.backends:108] (0.586)
## query1

SELECT "concordia_asset"."id", "concordia_asset"."item_id", "concordia_asset"."published", "concordia_asset"."title"
, "concordia_asset"."slug", "concordia_asset"."description", "concordia_asset"."media_url", "concordia_asset"."media_type"
, "concordia_asset"."sequence", "concordia_asset"."year", "concordia_asset"."resource_url", "concordia_asset"."download_url"
, "concordia_asset"."metadata", "concordia_asset"."transcription_status", "concordia_asset"."difficulty",
(SELECT U0."text"
 FROM "concordia_transcription" U0
 WHERE U0."asset_id" = ("concordia_asset"."id")
 ORDER BY U0."id" DESC  LIMIT 1) AS "latest_transcription"
 FROM "concordia_asset"
 INNER JOIN "concordia_item" ON ("concordia_asset"."item_id" = "concordia_item"."id")
 INNER JOIN "concordia_project" ON ("concordia_item"."project_id" = "concordia_project"."id")
 WHERE "concordia_asset"."item_id"
  IN (SELECT X0."id" FROM "concordia_item" X0
  INNER JOIN "concordia_project" X1 ON (X0."project_id" = X1."id")
  INNER JOIN "concordia_campaign" X2 ON (X1."campaign_id" = X2."id")
  WHERE (X2."slug" = 'garfield'
  AND NOT (X0."id"
   IN (SELECT W1."item_id" FROM "concordia_asset" W1
   WHERE W1."id" IN (SELECT V0."id"
   FROM "concordia_asset" V0
   WHERE (V0."item_id"
    IN (SELECT U0."id"
    FROM "concordia_item" U0
    INNER JOIN "concordia_project" U1 ON (U0."project_id" = U1."id")
    INNER JOIN "concordia_campaign" U2 ON (U1."campaign_id" = U2."id")
    WHERE U2."slug" = 'garfield')
    AND V0."transcription_status"
    IN ('not_started', 'in_progress', 'submitted')))))))
    ORDER BY "concordia_project"."title" ASC, "concordia_asset"."item_id" ASC, "concordia_asset"."sequence" ASC;
    args=('garfield', 'garfield', 'not_started', 'in_progress', 'submitted')

[2021-07-14T13:17:22 DEBUG django.db.backends:108] (100.177).   **<----!  long query!**

## query2 - pg_admin4 won't return rows..the session becomes idle and the connection is lost.

SELECT "concordia_item"."id", "concordia_item"."project_id", "concordia_item"."published"
, "concordia_item"."title", "concordia_item"."item_url", "concordia_item"."item_id"
, "concordia_item"."description", "concordia_item"."metadata", "concordia_item"."thumbnail_url"
FROM "concordia_item"
INNER JOIN "concordia_asset" ON ("concordia_item"."id" = "concordia_asset"."item_id")
WHERE "concordia_asset"."id"
 IN (SELECT Y0."id"
 FROM "concordia_asset" Y0
 WHERE Y0."item_id"
  IN (SELECT X0."id"
  FROM "concordia_item" X0
  INNER JOIN "concordia_project" X1 ON (X0."project_id" = X1."id")
  INNER JOIN "concordia_campaign" X2 ON (X1."campaign_id" = X2."id")
  WHERE (X2."slug" = 'garfield'
  AND NOT (X0."id"
   IN (SELECT W1."item_id"
   FROM "concordia_asset" W1
   WHERE W1."id"
    IN (SELECT V0."id"
    FROM "concordia_asset" V0
    WHERE (V0."item_id"
     IN (SELECT U0."id"
     FROM "concordia_item" U0
     INNER JOIN "concordia_project" U1 ON (U0."project_id" = U1."id")
     INNER JOIN "concordia_campaign" U2 ON (U1."campaign_id" = U2."id")
     WHERE U2."slug" = 'garfield')
     AND V0."transcription_status" IN ('not_started', 'in_progress', 'submitted'))))))));
     args=('garfield', 'garfield', 'not_started', 'in_progress', 'submitted')

### if I modified postgresql parms in session in pgadmin as follows...

SET SESSION idle_in_transaction_session_timeout = '5min';
SET SESSION tcp_keepalives_idle=300;
SET SESSION tcp_user_timeout = 300;
SET SESSION tcp_keepalives_count = 300;

### ...the error msg returned query2 in pg_admin is:
ERROR: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

If I run the second query with an interactive bash terminal in the db's container - 
the 2nd query is 'Killed'

This is the explain plan information: raw and links to graphical
Also https://explain.dalibo.com/plan/Ryi <- with BUFFER, verbose
https://explain.dalibo.com/plan/1XF
raw:
```
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-------------------------------
 Nested Loop  (cost=13728.65..21751.49 rows=10252 width=357) (actual time=444.557..637.072 rows=2462 loops=1)
   ->  Nested Loop  (cost=13728.36..18536.98 rows=10252 width=4) (actual time=444.504..550.558 rows=2462 loops=1)
         ->  HashAggregate  (cost=13727.94..13830.46 rows=10252 width=4) (actual time=444.440..464.677 rows=2462 loops=1)
               Group Key: y0.id
               Batches: 1  Memory Usage: 529kB
               ->  Nested Loop  (cost=9854.06..13702.31 rows=10252 width=4) (actual time=360.467..421.818 rows=2462 loops=1)
                     ->  HashAggregate  (cost=9853.64..9857.88 rows=424 width=4) (actual time=360.407..360.691 rows=3 loops=1)
                           Group Key: x0.id
                           Batches: 1  Memory Usage: 37kB
                           ->  Hash Join  (cost=8507.69..9852.58 rows=424 width=4) (actual time=257.340..360.603 rows=3 loops=1)
                                 Hash Cond: (x0.project_id = x1.id)
                                 ->  Seq Scan on concordia_item x0  (cost=8482.28..9783.36 rows=10604 width=8) (actual time=8.436..186.828 rows=21207 loops=1)
                                       Filter: (NOT (hashed SubPlan 1))
                                       SubPlan 1
                                         ->  Nested Loop  (cost=4866.31..8464.50 rows=7111 width=4) (actual time=8.332..8.478 rows=0 loops=1)
                                               ->  HashAggregate  (cost=4865.89..4937.00 rows=7111 width=4) (actual time=8.316..8.448 rows=0 loops=1)
                                                     Group Key: v0.id
                                                     Batches: 1  Memory Usage: 217kB
                                                     ->  Nested Loop  (cost=196.88..4848.11 rows=7111 width=4) (actual time=8.207..8.325 rows=0 loops=1)
                                                           ->  HashAggregate  (cost=196.46..204.94 rows=848 width=4) (actual time=0.472..0.592 rows=3 loops=1)
                                                                 Group Key: u0.id
                                                                 Batches: 1  Memory Usage: 49kB
                                                                 ->  Nested Loop  (cost=4.49..194.34 rows=848 width=4) (actual time=0.121..0.506 rows=3 loops=1)
                                                                       ->  Nested Loop  (cost=4.20..25.33 rows=7 width=4) (actual time=0.072..0.213 rows=3 loops=1)
                                                                             ->  Seq Scan on concordia_campaign u2  (cost=0.00..4.31 rows=1 width=4) (actual time=0.012..0.032 rows=1 loops=1)
                                                                                   Filter: ((slug)::text = 'garfield'::text)
                                                                                   Rows Removed by Filter: 24
                                                                             ->  Bitmap Heap Scan on concordia_project u1  (cost=4.20..20.95 rows=7 width=8) (actual time=0.036..0.084 rows=3 loops=1)
                                                                                   Recheck Cond: (campaign_id = u2.id)
                                                                                   Heap Blocks: exact=3
                                                                                   ->  Bitmap Index Scan on concordia_project_campaign_id_89bf7875  (cost=0.00..4.20 rows=7 width=0) (actual time=0.017..0.024 rows=3 loops=1)
                                                                                         Index Cond: (campaign_id = u2.id)
                                                                       ->  Index Scan using concordia_item_project_id_c1d67f28 on concordia_item u0  (cost=0.29..22.99 rows=115 width=8) (actual time=0.018..0.029 rows=1 loops=3)
                                                                             Index Cond: (project_id = u1.id)
                                                           ->  Index Scan using concordia_asset_item_id_e289371f on concordia_asset v0  (cost=0.42..5.26 rows=22 width=8) (actual time=2.545..2.547 rows=0 loops=3)
                                                                 Index Cond: (item_id = u0.id)
                                                                 Filter: ((transcription_status)::text = ANY ('{not_started,in_progress,submitted}'::text[]))
                                                                 Rows Removed by Filter: 821
                                               ->  Index Only Scan using concordia_a_id_137ca8_idx on concordia_asset w1  (cost=0.42..0.50 rows=1 width=8) (never executed)
                                                     Index Cond: (id = v0.id)
                                                     Heap Fetches: 0
                                 ->  Hash  (cost=25.33..25.33 rows=7 width=4) (actual time=0.239..0.301 rows=3 loops=1)
                                       Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                       ->  Nested Loop  (cost=4.20..25.33 rows=7 width=4) (actual time=0.090..0.238 rows=3 loops=1)
                                             ->  Seq Scan on concordia_campaign x2  (cost=0.00..4.31 rows=1 width=4) (actual time=0.020..0.041 rows=1 loops=1)
                                                   Filter: ((slug)::text = 'garfield'::text)
                                                   Rows Removed by Filter: 24
                                             ->  Bitmap Heap Scan on concordia_project x1  (cost=4.20..20.95 rows=7 width=8) (actual time=0.045..0.098 rows=3 loops=1)
                                                   Recheck Cond: (campaign_id = x2.id)
                                                   Heap Blocks: exact=3
                                                   ->  Bitmap Index Scan on concordia_project_campaign_id_89bf7875  (cost=0.00..4.20 rows=7 width=0) (actual time=0.020..0.027 rows=3 loops=1)
                                                         Index Cond: (campaign_id = x2.id)
                     ->  Index Scan using concordia_asset_item_id_e289371f on concordia_asset y0  (cost=0.42..8.45 rows=62 width=8) (actual time=0.021..7.534 rows=821 loops=3)
                           Index Cond: (item_id = x0.id)
         ->  Index Only Scan using concordia_a_id_137ca8_idx on concordia_asset  (cost=0.42..0.46 rows=1 width=8) (actual time=0.011..0.011 rows=1 loops=2462)
               Index Cond: (id = y0.id)
               Heap Fetches: 0
   ->  Index Scan using concordia_item_pkey on concordia_item  (cost=0.29..0.31 rows=1 width=357) (actual time=0.011..0.011 rows=1 loops=2462)
         Index Cond: (id = concordia_asset.item_id)
 Planning Time: 3.563 ms
 Execution Time: 657.168 ms
(61 rows)
```
Lots of nested loops. 

What I find interesting is that there are queries through the UI that take significantly more time - for example the initial loading of a campaign for transcription or the initial load and paging through assets in admin. 
The request is passed without mechanism to stay connected for the duration of the task. Or to signal back that to the UI that it's done ...and the file can be picked up? or downloaded.... ?

**Comment by jkueloc** [comment]
@elainekamlley  @lalgee I had already attached the production log from running Garfield export in production attached above - https://github.com/LibraryOfCongress/concordia/issues/1503#issuecomment-884246326

**Comment by rabiloc** [comment]
@lalgee As a temporary solution @jkueloc can get the latest production dump and create the bag it structure for all the campaigns and attach it to this ticket. We can tackle the bigger "timeout" issue progressively.

**Comment by lalgee** [comment]
Sounds good @rabiloc.  As noted above Dickinson has the error as well and also recently completed, but we are waiting on curator QC before we would need to do the export workaround.

**Comment by lalgee** [comment]
@jkueloc the exported bags in the AWS crowd-dev export directories are complete and correct.

**Comment by jkueloc** [comment]
@lalgee apologies for my confusion - is the note above meant for this ticket or for 500 error trying to export George Washington BagIt #1508?  I'm still missing something on this one based on testing with @laurenseroka last Friday.

**Comment by lalgee** [comment]
It's related to this ticket. You had checked if the campaigns timing out still  exported bags to that space and asked if the CM's could confirm that they are whole and complete. I think the thought was that AWS could be a workaround -- i.e. if an export times out we can just retrieve the data from AWS?

**Comment by lalgee** [comment]
This is also an issue for the Mary Church Terrell Campaign.

**Comment by jkueloc** [comment]
Linking another instance of the same base issue here that requires the local dev environment workaround  #1584 
Also capturing here that a decision was made after discussions of the base issue during meetings to delay work on issue since a workaround exits using a **local dev environment** .

**Comment by jkueloc** [comment]
This issue is closed as the work has been moved to the Library's Jira repo as noted above.
